### PR TITLE
Rspamd: add greylisting option & code refactoring

### LINKED ISSUE #3215: [FR] implement intelligent grey listing
# Feature Request

## Context
to reduce spam greylisting can be a very effective tool, however it causes many issues for the timely delivery of email and end users complain.

My own implementing of greylisting was very shortlived after receiving numerous complaints within hours of turning it on. I have not re-implemented greylisting since then, about 5 years ago.

The default delay is particularly a problem when getting security codes by email which have very short lifespans often measured in minutes. the user never receives the code even though they are hovering over their mail client waiting to pounce.

### Is your Feature Request related to a Problem?
Yes, spam reduction

### Describe the Solution you'd like
Can I suggest looking to implement intelligent grey listing, this means that the email is automatically passed through if extra checks pass such as valid SPF records

This active project seems to fit the bill and is already available for postfix
https://github.com/evgeni/bley

I am not suggesting it is the solution to use, I am not related to that project however it seems an excellent starting point.

### Are you going to implement it?

Yes, because I know the probability of someone else doing it is low and I can learn from it.

### What are you going to contribute??

Not sure what I can contribute but am happy to if I get it working properly and securely

## Additional context
None

### Alternatives you've considered
None at this stage

### Who will that Feature be useful to?
All users wishing to reduce spam but not at the loss of timely email delivery. so in short administrators of mail systems rather than individual users

### What have you done already?
I am just now implementing Docker-mailserver for the first time and was reviewing the existing greylisting tool as it implements first sight greylisting apparently irrespective of the reputation of the sender (SPF/DKIM) it is a blunt force tool.

As such I started looking for alternatives.

### PR DESCRIPTION
**Pull request by georglauterbach**
# Description

1. adds the option to easily enable greylisting
2. refactors setup code (no change in functionality)
3. changes order in which Redis & Rspamd are started - Redis now comes first (which makes more sense)

**I highly advise reviewing the first commit separately!** It is a pure refactoring.

Closes #3215

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] Improvement (non-breaking change that does improve existing functionality)
- [x] This change requires a documentation update

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (README.md or the documentation under `docs/`)
- [x] If necessary I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: mailserver.env [hunk 1]
```diff
@@ -125,6 +125,23 @@ ENABLE_RSPAMD=0
...
 ENABLE_RSPAMD_REDIS=
 
+# When enabled,
+#
+# 1. the "[autolearning][rspamd-autolearn]" feature is turned on;
+# 2. the Bayes classifier will be trained when moving mails from or to the Junk folder (with the help of Sieve scripts).
+#
+# **0** => disabled
+# 1     => enabled
+RSPAMD_LEARN=0
```
Comment (georglauterbach) [inline mailserver.env:135]: This should have been added with #3159, but I forgot about it..

**Comment by georglauterbach** [comment]
Just FYI: [Rspamd 3.5](https://github.com/rspamd/rspamd/blob/master/ChangeLog) landed on 17 Mar, so v12.0.0 will include the (currently) latest version of Rspamd 🥳🎉

This PR is intended for v12.1.0 though :)

File: target/scripts/startup/setup.d/security/rspamd.sh [hunk 2]
```diff
@@ -142,14 +149,74 @@ function __rspamd__disable_default_modules
...
+  fi
+}
+
+# Sets up greylisting based on the environment variable RSPAMD_GREYLISTING.
+function __rspamd__setup_greylisting
+{
+  if [[ ${RSPAMD_GREYLISTING} -eq 1 ]]
+  then
+    __rspamd__log 'debug' 'Enabling greylisting'
+    sed -i -E "s|(enabled =).*|\1 true;|g" /etc/rspamd/local.d/greylist.conf
```
Comment (georglauterbach) [inline target/scripts/startup/setup.d/security/rspamd.sh:210]: I guess we could use `sedfile` here, but not 100% sure 🤔

Comment (georglauterbach) [reply target/scripts/startup/setup.d/security/rspamd.sh]: Done in [3c6156e](https://github.com/docker-mailserver/docker-mailserver/pull/3206/commits/3c6156eb131a58d90d847938f7aebecea401c844)

**Comment by georglauterbach** [comment]
This PR is ready for review and merge now :)

**Comment by polarathene** [comment]
> PR is ready for review

At a glance there is nothing alarming, but I've not got the time to do a thorough review as is usually the case with the Rspamd feature work especially.

I did notice some concerns while collaborating on related docs, and I'll raise an issue about that for you to address separately.

---

I can provide an approval, but it won't mean too much :sweat_smile:

@casperklein do you want to provide a review, or should @georglauterbach go ahead and merge?

**Comment by georglauterbach** [comment]
@polarathene thanks for the info, much appreciated! Would you like to merge #3231 or this PR first?

**Comment by polarathene** [comment]
> first because the conflicts should be easier to resolve (if there are conflicts).

There will be conflicts with rspamd docs page. I agree that #3231 should be merged first.

**Comment by georglauterbach** [comment]
> > first because the conflicts should be easier to resolve (if there are conflicts).
> 
> 
> 
> There will be conflicts with rspamd docs page. I agree that #3231 should be merged first.

Go ahead and merge when you think the PR is good to go :)

**Comment by georglauterbach** [comment]
I marked this with `priority/high` now since it's blocking another PR I want to bring out :) I also resolved merge conflicts.

File: target/scripts/startup/setup.d/security/rspamd.sh [hunk 3]
```diff
@@ -31,33 +30,55 @@ function _setup_rspamd
...
+enabled = ${ENABLE_MODULE};
+
+EOF
+}
+
+# Run miscellaneous early setup tasks and checks, such as creating files needed at runtime
+# or checking for other anti-spam/anti-virus software.
+function __rspamd__run_early_setup_and_checks
+{
+  touch /var/lib/rspamd/stats.ucl
```
Comment (casperklein) [inline target/scripts/startup/setup.d/security/rspamd.sh:67]: In other places we use this:
```suggestion
  : >/var/lib/rspamd/stats.ucl
```

File: target/scripts/start-mailserver.sh [hunk 4]
```diff
@@ -129,8 +129,8 @@ function _register_functions
   [[ ${SMTP_ONLY}               -ne 1 ]] && _register_start_daemon '_start_daemon_dovecot'
 
   [[ ${ENABLE_UPDATE_CHECK}     -eq 1 ]] && _register_start_daemon '_start_daemon_update_check'
-  [[ ${ENABLE_RSPAMD}           -eq 1 ]] && _register_start_daemon '_start_daemon_rspamd'
   [[ ${ENABLE_RSPAMD_REDIS}     -eq 1 ]] && _register_start_daemon '_start_daemon_rspamd_redis'
+  [[ ${ENABLE_RSPAMD}           -eq 1 ]] && _register_start_daemon '_start_daemon_rspamd'
```
Comment (casperklein) [inline target/scripts/start-mailserver.sh:133]: Maybe add a comment, that RSPAM startup order matters?

**Review by casperklein — COMMENTED** [review]
LGTM. Just two minor things.

File: target/scripts/start-mailserver.sh (hunk 4 above)
Comment (georglauterbach) [reply target/scripts/start-mailserver.sh:136]: Done in [1ecdfe6](https://github.com/docker-mailserver/docker-mailserver/pull/3206/commits/1ecdfe60abc48ec11ef1238455aa7bac686dec14)

File: target/scripts/startup/setup.d/security/rspamd.sh (hunk 3 above)
Comment (georglauterbach) [reply target/scripts/startup/setup.d/security/rspamd.sh]: Done in [1ecdfe6](https://github.com/docker-mailserver/docker-mailserver/pull/3206/commits/1ecdfe60abc48ec11ef1238455aa7bac686dec14)

**Comment by georglauterbach** [comment]
> LGTM. Just two minor things.

❤️ I adjusted the PR according to the PR feedback and I'll be mergin this then :)

**Pull request merged**